### PR TITLE
Add indexing structure to accelerate lookups

### PR DESCRIPTION
**Pull request by anthony-murphy**

### CODE REVIEW

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 1]
```diff
@@ -416,6 +417,66 @@ const forwardPred = (ref: LocalReferencePosition): boolean =>
...
+	 */
+	private readonly seqOrdered = new DoublyLinkedList<ObliterateInfo>();
+
+	private readonly startOrdered = new SortedSegmentSet<LocalReferencePosition>();
+
+	constructor(private readonly mergeTree: MergeTree) {}
+
+	public setMinSeq(minSeq: number): void {
+		// eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+		while (!this.seqOrdered.empty && this.seqOrdered.first?.data.seq! <= minSeq) {
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:443]: this now does nothing if there are not obliterates, and it's at most O(n) for the number of obliterates

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 2]
```diff
@@ -416,6 +417,66 @@ const forwardPred = (ref: LocalReferencePosition): boolean =>
...
+		this.startOrdered.addOrUpdate(start);
+	}
+
+	public empty(): boolean {
+		return this.startOrdered.size === 0;
+	}
+
+	public findOverlapping(seg: ISegment): Iterable<ObliterateInfo> {
+		const overlapping: ObliterateInfo[] = [];
+		for (const start of this.startOrdered.items) {
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:465]: worst case scenario this is O(n) for the number of obliterates. we could probably accelerate this further with a tree.

File: packages/dds/merge-tree/src/mergeTree.ts (hunk 1 above)
Comment (jzaffiro) [reply packages/dds/merge-tree/src/mergeTree.ts:443]: If we're at the minSeq, does that mean that the obliterate should be considered out of the collabWindow? I had been assuming that it was the opposite (===minSeq means still inside collabWindow).

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts:443]: minSeq is the lowest seq all clients have seen, so if there is an obliterate at the minSeq it means it was seen by all clients, and therefore shouldn't remove anything more, as only clients that haven't see the obliterate should have their inserts removed.

**Review by jzaffiro — APPROVED** [review]

**Pull request merged**